`torch.pow` Add type promotion support and fix issue with __rpow__

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37098 `torch.pow` Add type promotion support and fix issue with __rpow__**


### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

Differential Revision: [D21686207](https://our.internmc.facebook.com/intern/diff/D21686207)

nairbv committed:
Fix issues with `pow` type promotion

Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`

Cherry-picked from another stack: https://github.com/pytorch/pytorch/pull/32582

[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Pow.h:
```diff
@@ -9,6 +9,48 @@ struct TensorIterator;
+template <class T,
+  typename std::enable_if<std::is_integral<T>::value && std::is_signed<T>::value, T>::type* = nullptr>
+static inline HOST_DEVICE T powi(T a, T b) {
+  if ( b < 0 ) {
```
In NumPy, attempting to raise an integer array to a negative power will throw an error:

```
a = np.array((2))
a ** -1
: ValueError: Integers to negative integer powers are not allowed.
```

nairbv replied:
we could change it, though it would be an additional BC-breaking change. On master `torch.tensor(1) ** torch.tensor(-1)` returns `tensor(1)`. I think if we decide to change that too it could go in a separate PR.

mruberry replied:
Separating them seems OK. Can we put a comment describing the desired behavior in this PR, then?

nairbv replied:
added a comment

mruberry commented on test/test_torch.py:
```diff
@@ -5604,10 +5604,11 @@ def to_np(value):
-                actual2 = actual.pow_(exponent)
-                self.assertEqual(actual, expected.to(actual), allow_inf=True)
-                self.assertEqual(actual2, expected.to(actual), allow_inf=True)
+                if torch.can_cast(torch.result_type(base, exponent), base.dtype):
```
@ngimel and I recommend running the other code path (that you can't cast) and asserting it throws an exception

nairbv replied:
added

mruberry commented on test/test_torch.py:
```diff
@@ -6092,6 +6093,7 @@ def test_pow(self, device):
                                                 r'Integers to negative integer powers are not allowed\.'):
                         torch.pow(m1[4], num)
                 else:
+                    expected_dtype = torch.result_type(num, m1[4])
```
Why m1[4]? @ngimel and I thought m1 is a tensor?

nairbv replied:
it's just the value that was being validated in existing tests (e.g. see above for `torch.pow(m1[4], num)`)

mruberry replied:
Weird, but seems like not an issue for this change?

mruberry commented on test/test_torch.py:
```diff
@@ -6100,6 +6102,12 @@ def test_pow(self, device):
                         res2[i] = math.pow(m1[4][i], num)
                     self.assertEqual(res1, res2)
 
+                    # scalar ** tensor to enforce correct handling of dtypes for __rpow__().
```
Do we ever check scalar ** tensor with negative values in the tensor, especially negative integer values?
It looks like this PR also changes how tensor ** tensor integer power works. Is there test coverage here for that?

nairbv replied:
m1 is a multi-dimensional tensor, so m1[4] is still a random tensor for each dtype (which presumably contains negative values). existing/older tests below have `scalar ** tensor`.

mruberry commented on aten/src/ATen/test/pow_test.cpp:
```diff
@@ -110,39 +110,53 @@ void assert_eq(T val, T act, T exp) {
   ASSERT_EQ(act, exp);
```
Are the tests in this file actually run in CI? We couldn't find the test output in several builds, including the ASAN build. If it's not run, should we continue to test this in C++ or port the tests to Python?

mruberry replied:
(One reason to move this test to Python, for example, would be to run on CPU and CUDA easily.)

nairbv replied:
most of the tests are in python. I had to update this one to get things to build, but I also generally prefer the python tests.

mruberry commented on aten/src/ATen/test/pow_test.cpp:
```diff
@@ -110,39 +110,53 @@ void assert_eq(T val, T act, T exp) {
   for (const auto pow : pows) {
+    if ( dtype == kInt && pow > std::numeric_limits<int>::max()) {
+      // value cannot be converted to type int without overflow
+      EXPECT_THROW(tensor.pow(pow), std::runtime_error);
```
Is this code path triggered (@ngimel and I are reading in order so maybe it is below)? We are surprised that we detect this case.

nairbv replied:
yes, if I get rid of the EXPECT_THROW then the test results in `"value cannot be converted to type int without overflow: 3037000499" thrown in the test body.`

mruberry commented on aten/src/ATen/test/pow_test.cpp:
```diff
@@ -225,35 +238,79 @@ void tensor_pow_tensor(const Vals vals, c10::ScalarType vals_dtype, Pows pows, c
+  for (const auto val : vals) {
+    // 1 has special checks below
+    if ( val != 1 && val != -1) {
+      ASSERT_EQ(native::powi(val, T(-4)), 0);
```
Seems like this line should be an exception if we following NumPy.

nairbv replied:
I think there's a comment elsewhere, but we don't follow numpy here (and this PR preserves the behavior on master for integral powers)

mruberry commented on aten/src/ATen/test/pow_test.cpp:
```diff
@@ -225,35 +238,79 @@ void tensor_pow_tensor(const Vals vals, c10::ScalarType vals_dtype, Pows pows, c
+  }
+  T neg1 = -1;
+  ASSERT_EQ(native::powi(neg1, T(0)), 1);
+  ASSERT_EQ(native::powi(neg1, T(-1)), -1);
```
These and the followings would throw exceptions, too.

mruberry commented on aten/src/ATen/test/pow_test.cpp:
```diff
@@ -225,35 +238,79 @@ void tensor_pow_tensor(const Vals vals, c10::ScalarType vals_dtype, Pows pows, c
+
+  T one = 1;
+  ASSERT_EQ(native::powi(one, T(0)), 1);
+  ASSERT_EQ(native::powi(one, T(-1)), 1);
```
More exceptions here.

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/Pow.cpp:
```diff
@@ -28,7 +28,8 @@ Tensor& pow_out(Tensor& result, const Tensor& base, Scalar exp) {
     result.resize_as_(base).copy_(base);
   } else {
-    auto iter = TensorIterator::unary_op(result, base,
+    auto common_dtype = at::result_type(base, exp);
```
I think we should update the above condition and also check for 0.0 and 1.0 cases for complex. this was a follow up from the PR merged with some complex related changes before. would be awesome if you could add that :D

nairbv replied:
you just mean removing the `!exp.isComplex()` ? changed it -- let me know if that's correct.

anjali411 replied:
no that would be correct when the exponent is complex. All I meant was that for complex exponents = 0. + 0. j and 1. + 0.j, we should do exactly what we do for the case when real exponent is 0.0 or 1.0

nairbv replied:
ah right. I think that will go better in another PR just since it's an unrelated change, I can stack it on this one.

anjali411 replied:
great! thanks :D

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

nairbv committed:
Update on "`torch.pow` Add type promotion support and fix issue with __rpow__"



### **Cherry-picked from another stack:**
Some code review already occurred here: https://github.com/pytorch/pytorch/pull/32582

### Summary:

Fixes: https://github.com/pytorch/pytorch/issues/32436

The issue caused incorrect handling of dtypes for scalar ** tensor.
e.g. before this change:
```
>>> 5.5 ** torch.ones(5, dtype=torch.int32)
tensor([5, 5, 5, 5, 5], dtype=torch.int32)
```
should return a float tensor.

Also fixes a number of incorrect cases:
 * tensors to negative powers were giving incorrect results (1 instead
    of 0 or error)
 * Behavior wasn't consistent between cuda/cpu
 * large_value ** 1 in some cases gave a result not equal
    to large_value because of truncation in conversion to double and back.

BC-breaking:

Previously incorrect behavior (in 1.4):
```
>>> a
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
>>> a.pow_(.5)
tensor([1, 1, 1, 1, 1], dtype=torch.int32)
```

After this change:
`RuntimeError: result type Float can't be cast to the desired output type Int`



[ghstack-poisoned]

mruberry reviewed (approved):
Awesome! Thanks for adding that comment. LGTM!
